cluster-provision/images/vm-image-builder/create-containerdisk.sh

```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env bash
+set -exuo pipefail
+
+SCRIPT_PATH=$(dirname "$(realpath "$0")")
+
+export CUSTOMIZE_IMAGE_SCRIPT=${CUSTOMIZE_IMAGE_SCRIPT:-"${SCRIPT_PATH}/customize-image.sh"}
+
+function customize_image() {
+  local source_image=$1
+  local os_variant=$2
+  local customized_image=$3
+  local cloud_config=$4
+
+  # Backup the VM image and pass copy of the original image
+  # in case customizing script fail.
+  vm_image_copy="$(dirname $customized_image)/copy-${source_image}"
+  cp "$source_image" "$vm_image_copy"
+
+  # TODO: convert this script and its dependencies to container
+  ${CUSTOMIZE_IMAGE_SCRIPT} "$vm_image_copy" "$os_variant" "$customized_image" "$cloud_config"
+
+  # Backup no longer needed.
+  rm -f "$vm_image_copy"
+}
+
+function cleanup() {
+    if [ $? -ne 0 ];then
+      rm -rf "${build_directory}"
+    fi
+
+    rm -f "copy-${VM_IMAGE}"
+}
+
+export IMAGE_NAME=${IMAGE_NAME:-example-fedora}
+export TAG=${TAG:-32}
+export OS_VARIANT=${OS_VARIANT:-fedora31}
+export CLOUD_CONFIG_PATH=${CLOUD_CONFIG_PATH:-"${SCRIPT_PATH}/example/cloud-config"}
+export VM_IMAGE_URL=${VM_IMAGE_URL:-$(cat "${SCRIPT_PATH}/example/image-url")}
+
+readonly VM_IMAGE="source-image.qcow2"
+readonly build_directory="${IMAGE_NAME}_build"
+readonly new_vm_image_name="provisioned-image.qcow2"
+
+trap 'cleanup' EXIT SIGINT
```

Comment:
just Exit is needed since it will run all the time

another thing is that -e might wont catch errors in functions according what read and saw once,
i would suggest to inject an error for example if VM_IMAGE doesnt exists 
and see if the cleanup is triggered
if it doesnt 
the -e should be replaced to -E and then it should catch it

Reply:
I will remove `SIGINT`

If VM_IMAGE is removed, since `-u` flag is up the script stops with `unbound variable` error
If `-u` is removed as well `cleanup` still invokes
I am not familiar with `-E` what it does?

Reply:
dont remove the var, remove the file the var point to.
trap might not catch errors that happen in functions even there is -e
-E should fix it if it indeed happen when u inject an error

Reply:
`VM_IMAGE` is path to downloaded image, if you set it to bad path it still works and cleanup is invoked
I saw in man page that `-E` works with trap that catch `ERR` interrupts, I dont think we need it in this script

Reply:
question is if the first error that happened was in a function, 
can you just add please `false` in a function and make sure it still runs indeed?

Reply:
I checked and it works

Reply:
thx